### PR TITLE
UniswapV3 VIP

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description

[TKR-46]

Official router for reference: https://github.com/Uniswap/uniswap-v3-periphery/blob/main/contracts/SwapRouter.sol

```
BUY ETH->USDC 0.02 -> 100.00 ($100.00) after 5.0s @ dev
	✔ PASS 0.02 -> 101.00 ($217.17) ($115.64)
	Uniswap_V3: 100%
	gas: 136615
BUY USDC->ETH 99.42 -> 0.02 ($100.00) after 5.0s @ dev
	✔ PASS 100.41 -> 0.02 ($190.35) ($89.94)
	Uniswap_V3: 100%
	gas: 106248
SELL USDC->WETH 100.00 -> 0.02 ($100.00) after 5.0s @ dev
	✔ PASS 100.00 -> 0.02 ($24.70) ($75.88)
	Uniswap_V3: 100%
	gas: 89648
SELL WETH->USDC 0.02 -> 99.48 ($100.00) after 5.0s @ dev
	✔ PASS 0.02 -> 99.48 ($15.45) ($84.03)
	Uniswap_V3: 100%
	gas: 99272
```

Here's a 2-hop:
```
BUY USDC->WETH 98.85 -> 0.02 ($100.00) after 5.0s @ dev
	✔ PASS 99.84 -> 0.02 ($100.57) ($0.73)
	Uniswap_V3: 100%
	gas: 177834
```

Here's a 2-hop that probably crosses tick boundaries:
```
BUY USDC->WETH 99605.04 -> 24.32 ($100000.00) after 5.0s @ dev
	✔ PASS 100545.79 -> 24.55 ($100548.01) ($2.23)
	Uniswap_V3: 100%
	gas: 541235
```

### TODO
- [x] simbot
- [x]  Write `UniswapV3Feature` tests

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    }
+
+    /// @dev The UniswapV3 pool swap callback which pays the funds requested
+    ///      by the caller/pool to the pool. Can only be called by a valid
+    ///      UniswapV3 pool.
+    /// @param amount0Delta Token0 amount owed.
+    /// @param amount1Delta Token1 amount owed.
+    /// @param data Arbitrary data forwarded from swap() caller. An ABI-encoded
+    ///        struct of: inputToken, outputToken, fee, payer
+    function uniswapV3SwapCallback(
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:180]: Gets called here: https://github.com/Uniswap/uniswap-v3-core/blob/main/contracts/UniswapV3Pool.sol#L776

**Review by mzhu25 — COMMENTED** [review]
Great stuff, left some style nits and one Multiplex consideration

File: packages/asset-swapper/contracts/src/BancorSampler.sol [hunk 2]
```diff
@@ -24,7 +24,7 @@ import "./interfaces/IBancor.sol";
 
 contract CompilerHack {}
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply packages/asset-swapper/contracts/src/BancorSampler.sol]: It'll be back

File: contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+    IUniswapV3Feature,
+    FixinCommon,
+    FixinTokenSpender
+{
+    /// @dev Name of this feature.
+    string public constant override FEATURE_NAME = "UniswapV3Feature";
+    /// @dev Version of this feature.
+    uint256 public immutable override FEATURE_VERSION = _encodeVersion(1, 0, 0);
+    /// @dev The deployed address of this contract.
+    address private immutable IMPLEMENTATION;
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:45]: Artifact from an experiment

File: contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        buyAmount = _swap(
+            encodedPath,
+            sellAmount,
+            minBuyAmount,
+            msg.sender,
+            address(this) // we are recipient because we need to unwrap WETH
+        );
+        WETH.withdraw(buyAmount);
+        // Transfer ETH to recipient.
+        _normalizeRecipient(recipient).call{ value: buyAmount }("");
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:145]: Doh

File: contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+                "UniswapV3Feature/INVALID_SWAP_CALLBACK_CALLER"
+            );
+        }
+        // Pay the amount owed to the pool.
+        if (amount0Delta > 0) {
+            _pay(token0, payer, msg.sender, uint256(amount0Delta));
+        } else if (amount1Delta > 0) {
+            _pay(token1, payer, msg.sender, uint256(amount1Delta));
+        } else {
+            revert("UniswapV3Feature/INVALID_SWAP_AMOUNTS");
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:217]: I don't think it works like V2. I tried prepaying the pool (hence the experiment) and the amounts were as if it were a fresh swap. I think you really have to pay in the callback

[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:217]: I can try again

[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:217]: oh fr? I'll reread their contracts

[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:217]: oh yeah you're right
less work for me then 😄

[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/features/UniswapV3Feature.sol:217]: yeah doesn't work

File: packages/asset-swapper/contracts/src/BancorSampler.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (mzhu25) [reply packages/asset-swapper/contracts/src/BancorSampler.sol]: 😨

**Review by mzhu25 — APPROVED** [review]

**Pull request merged**